array.h

```diff
@@ -2752,6 +2751,45 @@ void generate(array<T, Shape, Alloc>& dst, Generator&& g) {
   generate(dst.ref(), g);
 }
 
+/** Fill the `dst` array or array_ref with the result of calling a "pattern
+ * function" `fn`, which takes a single shape::index_type argument. The order in which `fn` is
+ * called is the same as `shape_traits<Shape>::for_each_index`. */
+template <typename T, typename Shape, class Fn>
+NDARRAY_HOST_DEVICE void transform_index(const array_ref<T, Shape>& dst, Fn&& fn) {
+  // In C++17 mode, support `transform_index` with rank 0. This is a rare corner case and more
```

Comment:
I'm surprised I haven't run into this before. I think maybe it's reasonable to have rank-0 helpers like this expect to take `fn(std::tuple<>)`?

I wonder if maybe it really should. Otherwise, writing rank-agnostic algorithms would be harder. In the case the algorithm isn't rank agnostic... this function really isn't necessary (it's replacing `dst() = fn()`).

Reply:
I'm suggesting just removing the special case, and letting rank 0 functions be `fn(std::tuple<>)` typed. That's the easy thing for rank-agnostic code, and for non-rank agnostic code, you can just write `dst() = fn()` which is clearer, simpler, and shorter.

Reply:
Got it. I approve. Done.